Signal/src/network/GiphyDownloader.swift

```diff
@@ -10,11 +10,79 @@ enum GiphyRequestPriority {
     case low, high
 }
 
+enum GiphyAssetSegmentState: UInt {
+    case waiting
+    case active
+    case complete
+    case failed
+}
+
+class GiphyAssetSegment {
+    let TAG = "[GiphyAssetSegment]"
+
+    public let index: UInt
+    public let segmentStart: UInt
+    public let segmentLength: UInt
+    public let redundantLength: UInt
+    public var state: GiphyAssetSegmentState = .waiting
+    private var datas = [Data]()
+
+    init(index: UInt,
+         segmentStart: UInt,
+         segmentLength: UInt,
+         redundantLength: UInt) {
+        self.index = index
+        self.segmentStart = segmentStart
+        self.segmentLength = segmentLength
+        self.redundantLength = redundantLength
+    }
+
+    public func totalDataSize() -> UInt {
+        var result: UInt = 0
+        for data in datas {
+            result += UInt(data.count)
+        }
+        return result
+    }
+
+    public func append(data: Data) {
+        datas.append(data)
+    }
+
+    public func mergeData(assetData: NSMutableData) {
+        // In some cases the last two segments will overlap.
+        // In that case, we only want to append the non-overlapping
+        // tail of the last segment.
+        var bytesToIgnore = Int(redundantLength)
+        for data in datas {
+            if data.count <= bytesToIgnore {
+                bytesToIgnore -= data.count
```

Comment:
I was being clever and trying to avoid doing redundant work:

* merging segment data
* merging asset data.

To avoid the first merge, I was storing a list of segment data pieces.  The code's a lot simpler if we just do two merges though so I've done that.